File: days/day2.swift
#!/usr/bin/env swift
import Foundation


func is_safe(_ data: [Int]) -> Bool {
    data[1...].enumerated().allSatisfy { p in [1,2,3].contains(p.element - data[p.offset]) }
    ||
    data[1...].enumerated().allSatisfy { p in [1,2,3].contains(data[p.offset] - p.element) }
}


func is_tolerable(_ data: [Int]) -> Bool {
    is_safe(data) ||
        !(0..<data.count)
            .allSatisfy { i in
                !is_safe(Array(data.prefix(upTo: i)) + Array(data.suffix(from: i+1)))
            }
}


func part1(_ data: String) -> Int {
    data
        .split(separator: "\n")
        .map { s in s.split(separator: " ").map { Int($0)! }}
        .map { is_safe($0) ? 1 : 0 }
        .reduce(0, +)
}


func part2(_ data: String) -> Int {
    data
        .split(separator: "\n")
        .map { s in s.split(separator: " ").map { Int($0)! }}
        .map { is_tolerable($0) ? 1 : 0 }
        .reduce(0, +)
}


let test1 = """
7 6 4 2 1
1 2 7 8 9
9 7 6 2 1
1 3 2 4 5
8 6 4 4 1
1 3 6 7 9
"""
assert(part1(test1) == 2)
assert(part2(test1) == 4)


let data = try! String(contentsOfFile: "day2.in", encoding: .utf8)
let ans1 = part1(data)
print("part1: \(ans1)")
let ans2 = part2(data)
print("part2: \(ans2)")

File: days/day2.jl
#!/usr/bin/env julia


function issafe(ar::AbstractVector)
    s = Set(diff(ar))
    return (s ⊆ [1, 2, 3]) || (s ⊆ [-1, -2, -3])
end


function part1(data)
    ans = 0
    for line in split(data, "\n", keepempty=false)
        xs = parse.(Int, split(line))
        ans += issafe(xs)
    end
    return ans
end


function part2(data)
    ans = 0
    for line in split(data, "\n", keepempty=false)
        xs = parse.(Int, split(line))
        for i in range(0, length(xs))
            if issafe([xs[1:i]; xs[i+2:end]])
                ans += 1
                break
            end
        end
    end
    return ans
end


data = """
7 6 4 2 1
1 2 7 8 9
9 7 6 2 1
1 3 2 4 5
8 6 4 4 1
1 3 6 7 9
"""

@assert part1(data) == 2
@assert part2(data) == 4


data = readchomp("day2.in")
println("part1: ", part1(data))
println("part2: ", part2(data))
